engine/crates/engine-v2/src/execution/planner/mod.rs

```diff
@@ -1,131 +1,217 @@
+        &self.build_context
+    }
+}
+
+impl<'ctx, 'op, R: Runtime> std::ops::DerefMut for ExecutionPlanner<'ctx, 'op, R> {
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        &mut self.build_context
+    }
+}
+
+// Ideally this BuilderContext would just be inside the ExecutionPlanner. But we do need to modify
+// the ExecutableOperation at some moments (here in the ExecutionPlanner) and at others we rely on it be immutable for walkers (in the builder::ExecutionBuilder)
+#[derive(Default)]
+pub struct BuildContext {
+    // Either an input or output field of a plan or response modifier
+    io_fields: Vec<FieldId>,
+    io_fields_buffer_pool: BufferPool<FieldId>,
+    response_modifier_executors: Vec<ResponseModifierExecutor>,
+    response_modifier_executors_input_fields: Vec<IdRange<IOFieldId>>,
+    response_modifier_executors_output_fields: Vec<IdRange<IOFieldId>>,
+    execution_plans: Vec<ExecutionPlan>,
+    execution_plans_input_fields: Vec<IdRange<IOFieldId>>,
     logical_plan_to_execution_plan_id: Vec<Option<ExecutionPlanId>>,
-    execution_plans_dependencies: Vec<Vec<FieldId>>,
-    response_view_selection_buffer_pool: BufferPool<ResponseViewSelection>,
     response_views: ResponseViews,
+    response_view_selection_buffer_pool: BufferPool<ResponseViewSelection>,
+}
+
+id_newtypes::NonZeroU16! {
+    BuildContext.io_fields[IOFieldId] => FieldId,
+}
+id_newtypes::index! {
+    ExecutionPlanner<'ctx, 'op, R: Runtime + 'static>.build_context.io_fields[IOFieldId] => FieldId,
+}
+
+impl BuildContext {
+    fn push_io_fields(&mut self, mut fields: Vec<FieldId>) -> IdRange<IOFieldId> {
+        let start = self.io_fields.len();
+        self.io_fields.extend(&mut fields.drain(..));
+        self.io_fields_buffer_pool.push(fields);
+        IdRange::from(start..self.io_fields.len())
+    }
+}
+
+pub(super) async fn plan<'ctx, R: Runtime>(
+    ctx: &PreExecutionContext<'ctx, R>,
+    prepared: Arc<PreparedOperation>,
+    variables: Variables,
+) -> PlanningResult<ExecutableOperation> {
+    let operation = ExecutableOperation {
+        query_modifications: query_modifier::QueryModificationsBuilder::new(ctx, &prepared, &variables)
+            .build()
+            .await?,
+        prepared,
+        variables,
+        subgraph_default_headers: create_subgraph_headers_with_rules(
+            ctx.request_context,
+            ctx.schema.walker().default_header_rules(),
+            http::HeaderMap::new(),
+        ),
+        execution_plans: Default::default(),
+        response_views: Default::default(),
+        response_modifier_executors: Default::default(),
+    };
+
+    let operation = ExecutionPlanner {
+        ctx,
+        build_context: BuildContext {
+            logical_plan_to_execution_plan_id: vec![None; operation.plan.logical_plans.len()],
+            io_fields: Vec::with_capacity(operation.fields.len()),
+            ..Default::default()
+        },
+        operation,
+    }
+    .build()?;
```

Comment:
`ExecutionPlanner::build` returns an operation?  I'd kind of assume that it returns a builder for or an instance of ExecutionPlanner.  Might be better named plan?  Although as I said above - if we could just get rid of this that'd be even better.